apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java

```diff
@@ -46,6 +53,28 @@ public class JmsMessageListenerInstrumentation extends BaseJmsInstrumentation {
     @SuppressWarnings("WeakerAccess")
     public static final Logger logger = LoggerFactory.getLogger(JmsMessageListenerInstrumentation.class);
 
+    @Nullable
+    private JmsConfiguration configuration;
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        if (configuration == null) {
+            configuration = GlobalTracer.getTracerImpl().getConfig(JmsConfiguration.class);
+        }
+
+        ElementMatcher.Junction<NamedElement> nameMatcher = nameContains("$") // inner classes
+            .or(nameContains("/")) // lambdas
+            .or(nameContains("Message"))
+            .or(nameContains("Listener"));
+
+        Collection<String> listenerPackages = configuration.getJmsListenerPackages();
+        if (listenerPackages.isEmpty()) {
+            return nameMatcher;
+        } else {
+            return nameMatcher.and(isInAnyPackage(listenerPackages, ElementMatchers.<NamedElement>none()));
+        }
```

Comment:
This can be a breaking change. We should document this very clearly and add to the CHANGELOG as such.

Reply:
:+1: in the case where it actually prevents instrumentation of a listener that was previously covered, do you think we should add now another internal option that would allow to completely bypass the naming check ? It might be a case where we'd better wait for feedback and know if it's actually breaking something before adding extra complexity.

Reply:
What I thought we can do is use EITHER the suffix prefilter OR the package-based prefilter. This way, you use these defaults and if someone defines packages, it overrides this, thus provides a workaround.
I think that if you provide a package where to look for a listener, it should be very quick and the other optimization is not necessarily required.
Just a thought.

Reply:
That's a very good idea !
- by default (`jms_listener_packages` is empty) we optimize with the new heuristic (and that makes a breaking change)
- when a non-empty value is set to `jms_listener_packages`, then we just use this package name as part of the heuristic, it would be up to the end-user to select a value that does not make the agent scan all classes.